### PR TITLE
 Implement iam server v5

### PR DESCRIPTION
**Pull request by mlohvynenko**

### CODE REVIEW

File: src/iamserver/registernodemanager.cpp [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+    auto ret = handler->HandleStream();
+
+    Remove(nodeID);
+
+    LOG_INF() << "Close register node stream. NodeID = " << nodeID.c_str() << " ret = " << ret.Message();
+
+    return ret.IsNone() ? grpc::Status::OK : grpc::Status(grpc::StatusCode::INTERNAL, ret.Message());
+}
+
+grpc::Status RegisterNodeManager::HandleSubscribeNodeChangedStream(
```
Comment (al1img) [inline src/iamserver/registernodemanager.cpp:307]: It should be separate class to handle node change subscriptions.

File: src/iamserver/registernodemanager.cpp [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+    grpc::ServerContext* context, grpc::ServerWriter<iamproto::NodeInfo>* writer)
+{
+    {
+        std::lock_guard lock(mNodeInfoWritersMutex);
+        mNodeInfoWriters.push_back(writer);
+    }
+
+    while (!context->IsCancelled()) {
```
Comment (al1img) [inline src/iamserver/registernodemanager.cpp:315]: And this looks as not proper implementation. Need to discuss.

File: src/iamserver/registernodemanager.cpp [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+
+grpc::Status RegisterNodeManager::HandleRegisterNodeStream(
+    const aos::NodeInfo& nodeInfo, RegisterNodeServerReaderWriter* stream, grpc::ServerContext* context)
+{
+    auto handler = std::make_shared<RegisterNodeStreamHandler>(stream, context);
+
+    const std::string nodeID = nodeInfo.mID.CStr();
+
+    if (!Add(nodeID, handler)) {
+        LOG_ERR() << "Close register node stream. NodeID has been already received: nodeID = " << nodeID.c_str();
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply src/iamserver/registernodemanager.cpp]: Refactored logic.

File: src/iamserver/registernodemanager.cpp [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+    do {
+        if (!mStream->Read(&outgoing)) {
+            return aos::ErrorEnum::eNone;
+        }
+
+        LOG_DBG() << "RegisterNodeStreamHandler received message: " << outgoing.DebugString().c_str();
+
+        std::lock_guard lock(mPendingMessagesMutex);
+
+        mPendingMessages[outgoing.IAMOutgoingMessage_case()].set_value(std::move(outgoing));
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply src/iamserver/registernodemanager.cpp]: I'd prefer to use move here `cause it simply swaps pointers on the underlying proto messages.

File: src/iamserver/publicmessagehandler.hpp [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+         * @param context server context.
+         * @param writer server writer.
+         * @return grpc::Status.
+         */
+        grpc::Status HandleStream(grpc::ServerContext* context, grpc::ServerWriter<T>* writer)
+        {
+            while (mIsRunning) {
+                if (context->IsCancelled()) {
+                    break;
+                }
```
Comment (MykolaSolyanko) [inline src/iamserver/publicmessagehandler.hpp:144]:
```
while(mRunning || !context->IsCancelled())
```

File: src/iamserver/publicmessagehandler.cpp [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+    mNodeChangedController.WriteToSteams(nodeInfo);
+}
+
+void PublicMessageHandler::OnNodeRemoved(const aos::String& id)
+{
+    (void)id;
+}
+
+aos::Error PublicMessageHandler::SubjectsChanged(const aos::Array<aos::StaticString<aos::cSubjectIDLen>>& messages)
+{
```
Comment (MykolaSolyanko) [inline src/iamserver/publicmessagehandler.cpp:79]: it will be good adding here a log

File: src/iamserver/publicmessagehandler.cpp [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+
+grpc::Status PublicMessageHandler::RegisterNode(grpc::ServerContext*                            context,
+    grpc::ServerReaderWriter<::iamproto::IAMIncomingMessages, ::iamproto::IAMOutgoingMessages>* stream)
+{
+    LOG_DBG() << "Public message handler. Process register node";
+
+    using aos::NodeStatus;
+    using aos::NodeStatusEnum;
+
+    static const std::vector<NodeStatus> allowedStatuses = {NodeStatusEnum::eUnprovisioned};
```
Comment (MykolaSolyanko) [inline src/iamserver/publicmessagehandler.cpp:362]: for my opinion will be better to move in const section, underhood lazy initialization used mutex

File: src/iamserver/protectedmessagehandler.cpp [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+
+grpc::Status ProtectedMessageHandler::RegisterNode(grpc::ServerContext*                         context,
+    grpc::ServerReaderWriter<::iamproto::IAMIncomingMessages, ::iamproto::IAMOutgoingMessages>* stream)
+{
+    LOG_DBG() << "Register node has been received on the protected channel";
+
+    using aos::NodeStatus;
+    using aos::NodeStatusEnum;
+
+    static const std::vector<NodeStatus> allowedStatuses = {NodeStatusEnum::eProvisioned, NodeStatusEnum::ePaused};
```
Comment (MykolaSolyanko) [inline src/iamserver/protectedmessagehandler.cpp:73]: for my opinion it could be better to move it in const section, and in theory it could be std::array instead of std::vector

File: src/iamserver/nodecontroller.cpp [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+aos::Error NodeStreamHandler::SendMessage(const iamproto::IAMIncomingMessages& request,
+    iamproto::IAMOutgoingMessages& response, const std::chrono::seconds responseTimeout)
+{
+    if (!mStream->Write(request)) {
+        return aos::ErrorEnum::eFailed;
```
Comment (al1img) [inline src/iamserver/nodecontroller.cpp:270]: Add error message and wrap here and in all the cases.

File: src/iamserver/nodecontroller.cpp [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error NodeStreamHandler::HandleNodeInfo(const iamproto::NodeInfo& info)
+{
+    aos::NodeInfo nodeInfo;
+
+    if (auto err = ConvertToAos(info, nodeInfo); !err.IsNone()) {
+        LOG_ERR() << "Failed to convert node info from proto to aos: " << err.Message();
```
Comment (al1img) [inline src/iamserver/nodecontroller.cpp:302]: Remove

File: src/iamserver/nodecontroller.cpp [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+
+        return AOS_ERROR_WRAP(err);
```
Comment (al1img) [inline src/iamserver/nodecontroller.cpp:304]: No need to wrap local errors.

File: src/iamserver/nodecontroller.cpp [hunk 12]
```diff
@@ -0,0 +1,493 @@
...
+    }
+
+    if (std::find(mAllowedStatuses.cbegin(), mAllowedStatuses.cend(), nodeInfo.mStatus) == mAllowedStatuses.cend()) {
+        LOG_ERR() << "Node status is not allowed: status = " << nodeInfo.mStatus.ToString();
```
Comment (al1img) [inline src/iamserver/nodecontroller.cpp:308]: remove spaces around =

File: src/iamserver/publicmessagehandler.cpp [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+{
+    mNodeController   = &nodeController;
+    mIdentHandler     = &identHandler;
+    mPermHandler      = &permHandler;
+    mNodeInfoProvider = &nodeInfoProvider;
+    mNodeManager      = &nodeManager;
+    mProvisionManager = &provisionManager;
+
+    if (auto err = mNodeInfoProvider->GetNodeInfo(mNodeInfo); !err.IsNone()) {
+        LOG_ERR() << "Failed to get node info: " << AOS_ERROR_WRAP(err);
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:37]: remove

File: src/iamserver/publicmessagehandler.cpp [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+            builder.RegisterService(static_cast<iamproto::IAMPublicIdentityService::Service*>(this));
+        }
+
+        builder.RegisterService(static_cast<iamproto::IAMPublicNodesService::Service*>(this));
+    }
+}
+
+void PublicMessageHandler::OnNodeInfoChange(const aos::NodeInfo& info)
+{
+    LOG_DBG() << "On node info changed: ID = " << info.mID;
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:65]: spaces around =

File: src/iamserver/publicmessagehandler.cpp [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+
+    iamproto::NodeInfo nodeInfo;
+    ConvertToProto(info, nodeInfo);
+
+    mNodeChangedController.WriteToSteams(nodeInfo);
+}
+
+void PublicMessageHandler::OnNodeRemoved(const aos::String& id)
+{
+    LOG_DBG() << "On node removed: ID = " << id;
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:75]: ditto in all the cases

File: src/iamserver/publicmessagehandler.cpp [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+}
+
+aos::Error PublicMessageHandler::SubjectsChanged(const aos::Array<aos::StaticString<aos::cSubjectIDLen>>& messages)
+{
+    LOG_DBG() << "Subjects changed: count = " << messages.Size();
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:80]: We are not so interested in count actually. Either log all subjects or remove count.

File: src/iamserver/publicmessagehandler.cpp [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+ **********************************************************************************************************************/
+
+/***********************************************************************************************************************
+ * IAMVersionService implementation
+ **********************************************************************************************************************/
+
+grpc::Status PublicMessageHandler::GetAPIVersion([[maybe_unused]] grpc::ServerContext* context,
+    [[maybe_unused]] const google::protobuf::Empty* request, iamanager::APIVersion* response)
+{
+    LOG_DBG() << "Public message handler. Process get API version";
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:148]: `Process get API version` is enough. Or add as additional parameter:
`Process get API version: handler=public`
or
`Process public get API version`

ditto for all the cases

File: src/iamserver/publicmessagehandler.cpp [hunk 18]
```diff
@@ -0,0 +1,440 @@
...
+/***********************************************************************************************************************
+ * IAMPublicIdentityService implementation
+ **********************************************************************************************************************/
+
+grpc::Status PublicMessageHandler::GetSystemInfo([[maybe_unused]] grpc::ServerContext* context,
+    [[maybe_unused]] const google::protobuf::Empty* request, iamproto::SystemInfo* response)
+{
+    LOG_DBG() << "Public message handler. Process get system info";
+
+    auto [systemID, err1] = GetIdentHandler()->GetSystemID();
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:212]: err1, err2 looks not so beaty. Define err on top.

File: src/iamserver/publicmessagehandler.cpp [hunk 19]
```diff
@@ -0,0 +1,440 @@
...
+
+    if (!err1.IsNone()) {
+        LOG_DBG() << "Get system ID error: " << err1;
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:215]: Remove

File: src/iamserver/publicmessagehandler.cpp [hunk 20]
```diff
@@ -0,0 +1,440 @@
...
+
+        return grpc::Status(grpc::StatusCode::INTERNAL, "Get system ID error");
+    }
+
+    auto [boardModel, err2] = GetIdentHandler()->GetUnitModel();
+    if (!err2.IsNone()) {
+        LOG_DBG() << "Get board model error: " << err2;
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:222]: LOG_ERR + error=

File: src/iamserver/publicmessagehandler.cpp [hunk 21]
```diff
@@ -0,0 +1,440 @@
...
+grpc::Status PublicMessageHandler::GetSubjects([[maybe_unused]] grpc::ServerContext* context,
+    [[maybe_unused]] const google::protobuf::Empty* request, iamproto::Subjects* response)
+{
+    LOG_DBG() << "Public message handler. Process get subjects";
+
+    aos::StaticArray<aos::StaticString<aos::cSubjectIDLen>, aos::cMaxSubjectIDSize> subjects;
+
+    auto err = GetIdentHandler()->GetSubjects(subjects);
+    if (!err.IsNone()) {
+        LOG_DBG() << "Get subjects error: " << err;
```
Comment (al1img) [inline src/iamserver/publicmessagehandler.cpp:242]: ditto

File: src/iamserver/nodecontroller.cpp [hunk 22]
```diff
@@ -9,108 +9,7 @@
 
 #include "log.hpp"
 #include "nodecontroller.hpp"
-
-/***********************************************************************************************************************
- * Static
- **********************************************************************************************************************/
-
-static aos::Error ConvertToAos(
```
Comment (al1img) [inline src/iamserver/nodecontroller.cpp]: It should be done as separate commit as add proto conversion to utils. And then use it in nodecontroller.

File: src/iamserver/nodecontroller.cpp [hunk 23]
```diff
@@ -0,0 +1,489 @@
...
+
+grpc::Status NodeController::HandleRegisterNodeStream(const std::vector<aos::NodeStatus>& allowedStatuses,
+    NodeServerReaderWriter* stream, grpc::ServerContext* context, aos::iam::nodemanager::NodeManagerItf* nodeManager)
+{
+    auto handler = std::make_shared<NodeStreamHandler>(allowedStatuses, stream, context, nodeManager);
+    StoreHandler(handler);
+
+    auto ret = handler->HandleStream();
+
+    handler->Close();
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply src/iamserver/nodecontroller.cpp:356]: Add atomic_bool to check if stream is closed.

File: src/iamserver/publicmessagehandler.hpp [hunk 24]
```diff
@@ -0,0 +1,224 @@
...
+         * @return grpc::Status.
+         */
+        grpc::Status HandleStream(grpc::ServerContext* context, grpc::ServerWriter<T>* writer)
+        {
+            while (mIsRunning && !context->IsCancelled()) {
+                std::shared_lock lock {mMutex};
+
+                if (mCV.wait_for(lock, cWaitTimeout, [this] { return mLastMessage.has_value(); })) {
+                    // got notification, send it to the client
+                    if (!writer->Write(*mLastMessage)) {
```
Comment (MykolaSolyanko) [inline src/iamserver/publicmessagehandler.hpp:147]: should we reset `mLastMessage` after we write a message? if it optional then wait_for will always pass, but in this case we will send the same message

**Review by MykolaSolyanko — APPROVED** [review]

File: src/iamserver/nodecontroller.cpp [hunk 25]
```diff
@@ -0,0 +1,399 @@
...
+    while (mStream->Read(&outgoing)) {
+        LOG_DBG() << "Receive message: type=" << outgoing.GetTypeName().c_str();
+
+        const auto messageCase = outgoing.IAMOutgoingMessage_case();
+        if (messageCase == iamproto::IAMOutgoingMessages::IAMOutgoingMessageCase::IAMOUTGOINGMESSAGE_NOT_SET) {
+            continue;
+        }
+
+        if (outgoing.has_node_info()) {
+            if (auto err = HandleNodeInfo(outgoing.node_info()); !err.IsNone()) {
```
Comment (al1img) [inline src/iamserver/nodecontroller.cpp:72]: You shouldn't break the loop in this case I guess. It is internal error not communication error.

File: src/iamserver/nodecontroller.cpp [hunk 26]
```diff
@@ -0,0 +1,399 @@
...
+                err = AOS_ERROR_WRAP(err);
+
+                break;
+            }
+        }
+
+        std::lock_guard lock {mMutex};
+
+        try {
+            mPendingMessages[messageCase].set_value(std::move(outgoing));
```
Comment (al1img) [inline src/iamserver/nodecontroller.cpp:82]: You should check if the message exists in `mPendingMessages` and skip them.

File: src/iamserver/nodecontroller.cpp (hunk 25 above)
Comment (mlohvynenko) [reply src/iamserver/nodecontroller.cpp:72]: Error is possible if secondary node is provisioned but connected to public IAM main (or uprovisioned to protected)
Also error is generated if protobuf conversion to AOS fails
Or node manager fails

Looks like we should notify secondary IAM somehow that error occurred.

File: src/iamserver/publicmessagehandler.hpp [hunk 27]
```diff
@@ -0,0 +1,224 @@
...
+         * @param context server context.
+         * @param writer server writer.
+         * @return grpc::Status.
+         */
+        grpc::Status HandleStream(grpc::ServerContext* context, grpc::ServerWriter<T>* writer)
+        {
+            while (mIsRunning && !context->IsCancelled()) {
+                std::shared_lock lock {mMutex};
+
+                if (mCV.wait_for(lock, cWaitTimeout, [this] { return mLastMessage.has_value(); })) {
```
[the remark this answers is not in the document]
Comment (mlohvynenko) [reply src/iamserver/publicmessagehandler.hpp:145]: I can't reset optional in this implementation on the write complete
But I've updated logic to assign determine if the message is new for a particular writer thread

File: src/iamserver/publicmessagehandler.hpp [hunk 28]
```diff
@@ -0,0 +1,233 @@
...
+
+            mCV.notify_all();
+        }
+
+        /**
+         * Writes notification message to all streams.
+         *
+         * @param message notification message.
+         */
+        void WriteToSteams(const T& message)
```
Comment (mykola-kobets-epam) [inline src/iamserver/publicmessagehandler.hpp:122]: streams*

File: src/iamserver/nodecontroller.hpp [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+
+    /**
+     * Destructor.
+     */
+    ~NodeStreamHandler();
+
+    /**
+     * Closes stream handler.
+     */
+    void Close();
```
Comment (mykola-kobets-epam) [inline src/iamserver/nodecontroller.hpp:52]: It's not clear why Close method required, destructor should be sufficient.

File: src/iamserver/protectedmessagehandler.hpp [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+#include <aos/iam/permhandler.hpp>
+#include <config/config.hpp>
+
+#include <iamanager/v5/iamanager.grpc.pb.h>
+
+#include "nodecontroller.hpp"
+#include "publicmessagehandler.hpp"
+
+/**
+ * Protected message handler. Responsible for handling public IAM services.
```
Comment (mykola-kobets-epam) [inline src/iamserver/protectedmessagehandler.hpp:30]: "for handling protected* IAM services"

File: tests/iamserver/iamserver_test.cpp [hunk 31]
```diff
@@ -252,739 +255,200 @@ void IAMServerTest::ApplyCertificate(const String& certType, const String& subje
...
-    SystemInfo              response;
-
-    EXPECT_CALL(mIdentHandler, GetSystemID)
-        .WillOnce(Return(aos::RetWithError<aos::StaticString<aos::cSystemIDLen>>(cSystemID)));
-    EXPECT_CALL(mIdentHandler, GetUnitModel)
-        .WillOnce(Return(aos::RetWithError<aos::StaticString<aos::cUnitModelLen>>("", aos::ErrorEnum::eFailed)));
-
-    const auto status = clientStub->GetSystemInfo(&context, request, &response);
-    ASSERT_FALSE(status.ok()) << status.error_message() << " (" << status.error_code() << ")";
+    EXPECT_EQ(status.error_code(), grpc::StatusCode::UNIMPLEMENTED)
```
Comment (mykola-kobets-epam) [inline tests/iamserver/iamserver_test.cpp:329]: I think iamserver_test.cpp should be removed, anyway test server doesn't implement anything here

**Review by mykola-kobets-epam — APPROVED** [review]

File: src/iamserver/nodecontroller.hpp (hunk 29 above)
Comment (mlohvynenko) [reply src/iamserver/nodecontroller.hpp:52]: handlers are shared_ptr(s).
Server code closes handler explicitly if stream handling fails.

File: tests/iamserver/iamserver_test.cpp (hunk 31 above)
Comment (mlohvynenko) [reply tests/iamserver/iamserver_test.cpp:329]: I'd keep it
This UT checks iam server init
Provision flag driven logic
and availability of IAM public services on the protected port

File: src/iamserver/nodecontroller.hpp (hunk 29 above)
Comment (mlohvynenko) [reply src/iamserver/nodecontroller.hpp:52]: The main intent is to make sure stream handler is properly closed when RPC closes
Otherwise, there might be an issue with reader-writer & context pointer

Comment (mykola-kobets-epam) [reply src/iamserver/nodecontroller.hpp:52]: and what is a reason not to reset shared_ptr instead?

File: tests/iamserver/iamserver_test.cpp (hunk 31 above)
Comment (mykola-kobets-epam) [reply tests/iamserver/iamserver_test.cpp:329]: so, basically it checks Init method only.
So I would suggest keep unit tests for Init method only and remove all the rest.

File: src/iamserver/nodecontroller.hpp (hunk 29 above)
Comment (mykola-kobets-epam) [reply src/iamserver/nodecontroller.hpp:52]: Ok, it looks like Close can be called multiple times. If so, I'm fine with that. 
If not, I don't see any reason to have both: destructor & Close.
